c10/core/Contiguity.h

```diff
@@ -12,49 +12,86 @@ namespace c10 {
 
 template <typename T>
 bool _compute_contiguous(ArrayRef<T> sizes, ArrayRef<T> strides, T numel) {
-  if (TORCH_GUARD_SIZE_OBLIVIOUS(sym_eq(numel, 0))) {
+  if (numel == 0) {
     return true;
   }
 
   T expected_stride = 1;
   // NB: make sure we do signed arithmetic
   for (int64_t d = int64_t(sizes.size()) - 1; d >= 0; d--) {
     const auto& size_d = sizes[d];
-    if (TORCH_GUARD_SIZE_OBLIVIOUS(sym_eq(size_d, 1))) {
+    if (size_d == 1) {
       continue;
     }
 
-    if (TORCH_GUARD_SIZE_OBLIVIOUS(sym_ne(strides[d], expected_stride))) {
+    if (strides[d] != expected_stride) {
       return false;
     }
     expected_stride *= size_d;
   }
   return true;
 }
 
-// This function will return True if the tensor is contiguous, and False if the
-// its not or if we can't determine if it is contiguous due to unbacked symbols
-// (it could be either in that case based on the actual runtime data).
-template <typename T>
-bool definitely_contiguous(ArrayRef<T> sizes, ArrayRef<T> strides, T numel) {
-  if (TORCH_GUARD_OR_FALSE(sym_eq(numel, 0))) {
+// Return a SymBool with underlying symbolic expression that represents
+// contiguity. Guaranteed not to add guards.
+inline static c10::SymBool _compute_contiguous_sym(
+    ArrayRef<c10::SymInt> sizes,
+    ArrayRef<c10::SymInt> strides,
+    const c10::SymInt& numel) {
+  // If this return true, the tensor is contiguous indeed. Otherwise it could be
+  // either.
+  auto is_contiguous_or_false = [&]() {
+    if (TORCH_GUARD_OR_FALSE(sym_eq(numel, 0))) {
+      return true;
+    }
```

Comment:
One subtlety, which I am not sure if it matters in practice: if the numel actually is zero, we will issue a number of guards on strides/sizes from below, even though all of it is irrelevant and actually you just need to test that numel is zero (contiguous!) In other words, it's not "free" to run the code below because it may issue guards that the original code wouldn't have issued. I'm willing to be persuaded from testing this doesn't matter though.

Reply:
1) if we can statically tell or if with adding guarding determine that numel ==0 then this function will exit with true, and  _compute_contiguous_sym will return True without adding  more guards.

2) I agree that if is_contiguous_or_false can add some guards that are not needed when is_contiguous_or_false return false, my mental model for this, if the user want a 100% guarantee for no guarding they should delete the hint.

I will remove the comment "Guaranteed not to add guards" its wrong i will add "Guaranteed not to throw data dependent errrors"